toolchain/sem_ir/cpp_file.cpp

```diff
@@ -0,0 +1,19 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "toolchain/sem_ir/cpp_file.h"
+
+namespace Carbon::SemIR {
+
+auto CppFile::VisitLocalTopLevelDecls(
+    llvm::function_ref<void(const clang::Decl*)> visitor) const -> void {
+  ast_unit_->visitLocalTopLevelDecls(
+      &visitor, [](void* erased_visitor_ptr, const clang::Decl* decl) {
+        auto* visitor_ptr = static_cast<decltype(visitor)*>(erased_visitor_ptr);
+        (*visitor_ptr)(decl);
```

Comment:
Would it make sense to pass differently to avoid the cast:

```suggestion
  ast_unit_->visitLocalTopLevelDecls(
      /*context=*/nullptr, [&](void* unused_context, const clang::Decl* decl) {
        visitor(decl);
```

Reply:
That'd be nice, but `visitLocalTopLevelDecls` takes a function pointer not a `function_ref`, so we can't pass a lambda with captures.

Reply:
FWIW if there's a cast that could be done on the function_ref instead of the void*, that would seem nicer since I assume it'd get more type checking.

Reply:
I don't know of a better way of expressing this. Would be happy if you have a suggestion, but otherwise I guess we hold our noses and apply this. I'd be tempted to push a change into upstream Clang to support a `function_ref` here, but I'm also thinking about switching away from `ASTUnit` which would resolve this too.

Reply:
Ah, I see, so can't even pass a capturing lambda here. I'd thought it might still be possible, just wrong about that.